### PR TITLE
My double-down attempt to upgrade OtlContainers.pas

### PR DESCRIPTION
**Pull request by SeanBDurkin**

### CODE REVIEW

**Comment by SeanBDurkin** [comment]
Oops , I thought I already did this. Trying again.

File: OtlTaskControl.pas [hunk 1]
```diff
@@ -413,7 +418,7 @@   TOmniMessageExec = class
     function  Join(const group: IOmniTaskGroup): IOmniTaskControl;
     function  Leave(const group: IOmniTaskGroup): IOmniTaskControl;
     function  MonitorWith(const monitor: IOmniTaskControlMonitor): IOmniTaskControl;
-    function  MsgWait(wakeMask: DWORD = QS_ALLEVENTS): IOmniTaskControl;
```
Comment (gabr42) [inline OtlTaskControl.pas]: Default parameter has to stay, at least in MSWINDOWS.

File: OtlSync.pas [hunk 2]
```diff
@@ -380,6 +367,22 @@     TAutoUnlock = class(TInterfacedObject, IOmniLockManagerAutoUnlock)
     function  BaseEvent: TEvent;
   end;
 
+  IOmniCancellationToken = interface ['{5946F4E8-45C0-4E44-96AB-DBE2BE66A701}']
```
Comment (gabr42) [inline OtlSync.pas]: IOmniCancellationToken must exist in Delphi 2007 (where OTL_Generics is not defined).

File: OtlComm.pas [hunk 3]
```diff
@@ -135,54 +137,45 @@   TOmniMessageQueue = class;
...
-    property EventObserver: TOmniContainerWindowsEventObserver read mqWinEventObserver;
+    function  GetNewMessageEvent: IOmniEvent;
+    function  TryDequeue(var msg: TOmniMessage): boolean;
+    property EventObserver: TOmniContainerEventObserver read mqEventObserver;
+   {$IFDEF MSWINDOWS}
     property OnMessage: TOmniMessageQueueMessageEvent read mqWinMsgObserver.OnMessage write SetOnMessage;
+    {$ENDIF}
   end; { TOmniMessageQueue }
 
-  IOmniMessageQueueTee = interface ['{8A9526BF-71AA-4D78-BAE8-3490C3987327}']
```
Comment (gabr42) [inline OtlComm.pas]: My code uses this interface+class, don't remove it!

File: OtlSync.pas [hunk 4]
```diff
@@ -226,19 +226,6 @@   TOmniResourceCount = class(TInterfacedObject, IOmniResourceCount)
   end; { TOmniResourceCount }
   {$ENDIF}
 
-  IOmniCancellationToken = interface ['{5946F4E8-45C0-4E44-96AB-DBE2BE66A701}']
```
Comment (gabr42) [inline OtlSync.pas]: Should stay outside of IFDEF OTL_Generics.

File: OtlTaskControl.pas [hunk 5]
```diff
@@ -966,7 +985,7 @@   TOmniTaskControl = class(TInterfacedObject, IOmniTaskControl, IOmniTaskControl
...
```
Comment (gabr42) [inline OtlTaskControl.pas]: Default value should stay, at least in MSWINDOWS.

File: OtlTaskControl.pas [hunk 6]
```diff
@@ -1656,18 +1682,52 @@ destructor TOmniTaskExecutor.Destroy;
   FreeAndNil(oteTerminateHandles);
   FreeAndNil(oteMethodHash);
   FreeAndNil(oteException);
-  DSiCloseHandleAndNull(oteCommRebuildHandles);
```
Comment (gabr42) [inline OtlTaskControl.pas]: Where did those two lines go?????

File: OtlTaskControl.pas [hunk 7]
```diff
@@ -2425,13 +2487,23 @@ function TOmniTaskExecutor.WaitForInit: boolean;
   Result := WorkerInitOK;
 end; { TOmniTaskExecutor.WaitForInit }
 
-function TOmniTaskExecutor.WaitForEvent(msgInfo: TOmniMessageInfo; timeout_ms: cardinal):
```
Comment (gabr42) [inline OtlTaskControl.pas]: Please rebase to the latest version before creating new pull request.

File: OtlSync.pas (hunk 2 above)
Comment (SeanBDurkin) [inline OtlSync.pas]: IOmniCancellationToken is on line 370.

File: OtlSync.pas (hunk 4 above)
Comment (SeanBDurkin) [inline OtlSync.pas]: This bit I do not understand. It already is outside of OTL_Generics

File: OtlTaskControl.pas (hunk 1 above)
Comment (SeanBDurkin) [inline OtlTaskControl.pas]: Done.

File: OtlTaskControl.pas (hunk 5 above)
Comment (SeanBDurkin) [inline OtlTaskControl.pas]: Done.

File: OtlTaskControl.pas (hunk 6 above)
Comment (SeanBDurkin) [inline OtlTaskControl.pas]: It is still there. It is just hidden by compiler magic. oteCommRebuildHandles is now an interface pointer of type IOmniEvent. When the TOmniTaskExecutor object is destroyed, the interface data members are de-referenced, and when the reference count reaches zero, the event resource is then de-allocated.

Comment (gabr42) [inline OtlTaskControl.pas]: Thanks.

**Pull request closed without merging**